### ISSUE TITLE
jsonrpcclient.exceptions.ReceivedErrorResponseError: Internal Error

### ISSUE THREAD

**Issue by ccgauvin94** [nextpyp/pyp#2]
Apologies if I've got the wrong log here. But I'm running into an issue I can't quite grok. I'm using an auto-scaling SLURM cluster, with a persistent login node, but workers that don't exist until a job is run. 

I can make a project, and submit an Import job for 9 micrographs. The weird things happen with the first job - it never actually spins up the worker. If I re-launch the import job (or just make a 2nd import job), it does in fact start the worker. The job log shows that the job runs OK, and then fails at the very end:

```
ontconfig error: No writable cache directories
2024-08-13 20:19:07 [INFO] /opt/pyp/bin/run/pyp:3232 | nextPYP v0.6.5 launching job 2 on pipeline-20519 using 1 task(s) and 4 GB of RAM
2024-08-13 20:19:12 [INFO] /opt/pyp/bin/run/pyp:380 | Attempting to load existing parameters and metadata from /cryoem_share/<redacted>_V1722941820_06262024/data/smaller_subset
2024-08-13 20:19:18 [INFO] /opt/pyp/bin/run/pyp:422 | /cryoem_share/<redacted>_V1722941820_06262024/data/smaller_subset/*.tif found 9 files to link into raw folder
2024-08-13 20:19:24 [INFO] /opt/pyp/bin/run/pyp:4098 | Selecting image for preview: /cryoem_share/<redacted>_V1722941820_06262024/data/smaller_subset/20240614_<redacted>-V1722941820_1_Jun1
5_22.00.39.tif
2024-08-13 20:19:31 [INFO] /opt/pyp/bin/run/pyp:4101 | Image dimensions are 5760 x 4092 (37 frames/tilts)
2024-08-13 20:20:16 [INFO] /opt/pyp/bin/run/pyp:4123 | PYP (pypgain) finished successfully
Traceback (most recent call last):
  File "/opt/pyp/bin/run/webrpc", line 47, in <module>
    cli()
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/opt/pyp/bin/run/webrpc", line 23, in slurm_ended
    Web().slurm_ended(_arrayid(), exit)
  File "/opt/pyp/src/pyp/streampyp/web.py", line 78, in slurm_ended
    self._request(Request.slurm_ended(webid=self.webid, arrayid=arrayid, exit_code=exit_code))
  File "/opt/pyp/src/pyp/streampyp/web.py", line 39, in _request
    response = self.client.send(request)
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/apply_defaults/decorators.py", line 13, in wrapper
    return function(self, *args, **kwargs)
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/jsonrpcclient/client.py", line 177, in send
    raise ReceivedErrorResponseError(response.data)
jsonrpcclient.exceptions.ReceivedErrorResponseError: Internal Error
```

I seem to get the same issue if I try to run the ping command. 

Here is my config: 

```
[pyp]
# path to the PYP container file, from a SLURM compute node's point of view
container = '/opt/nextPYP/pyp.sif'
# fast storage for temporary files, ideally local to the compute node
scratch = '/common/nextpyp_scratch'
binds = [ '/cryoem_share' ]
[web]
port = 8081
debug = true
localDir = '/cryoem_share/nextpyp/local'
sharedDir = '/common/nextpyp_shared'
workflowDirs = ['/opt/nextPYP/workflows']
auth = "none"
host = '0.0.0.0'
webhost = 'http://pipeline-20518:8081'
[slurm]
host = 'pipeline-20518'
user = 'nextpyp'
path = '/usr/bin'
queues = ['main.q']
```

I've verified that: 

1) The Docker containers are being launched when (all but the 1st) jobs are submitted
2) The spun-up containers can resolve the slurm host to the master node
3) The spun-up containers can SSH into the master node without a password
4) The filesystems are all in the same locations on the slurm worker as on the master, and with the same r/w permissions

Unfortunately, the error `jsonrpcclient.exceptions.ReceivedErrorResponseError: Internal Error` is too vague for me to figure out what could be causing it. Is there anywhere else to look for log output? Unfortunately, the micromon log doesn't show much: 

```
[2024-08-12 22:37:14,829]-[main] INFO  Backend - MicroMon started!
[2024-08-12 22:37:14,833]-[main] INFO  Backend - CWD: /opt/micromon
[2024-08-12 22:37:15,002]-[main] INFO  Backend - Read configuration from: /opt/nextPYP/config.toml
[2024-08-12 22:37:15,205]-[main] INFO  Backend - Configuration:
[]
               dev: false
[pyp]
       conatainer:  /opt/nextPYP/pyp.sif
          sources:  null
          scratch:  /tmp/pyp
            binds:  
         cudaLibs:  
[standalone]
     available cpus:  3
   available memory:  12 GiB
     available gpus:  1
[web]
             host:  127.0.0.1
             port:  8080
        local dir:  /opt/nextPYP/local
       shared dir:  /opt/nextPYP/shared
             auth:  None
          webhost:  http://127.0.0.1:8080
            debug:  false
     sendmail cmd:  null
         heap MiB:  2048
      database GB:  1.0
              JMX:  false
   OOM heap dumps:  false
    workflow dirs:  /opt/nextPYP/workflows
        demo mode:  false
        runas dir:  /opt/nextPYP/shared/runas
Starting Netty engine with these settings:
    Number of available processors:  4
            Connection thread pool:  3 threads
                Worker thread pool:  3 threads
               Request thread pool:  4 threads
   Worker and Request pools shared:  false
               Request queue limit:  16
         Request concurrency limit:  10
            Response write timeout:  10 s
              Request read timeout:  0 s
[2024-08-12 22:37:16,216]-[main] INFO  ktor.application - Autoreload is disabled because the development mode is off.
[2024-08-12 22:37:17,985]-[main] INFO  Migrations - No migrations to run. Everything up to date.
[2024-08-12 22:37:17,997]-[main] INFO  Backend - Reading workflow: /opt/nextPYP/workflows/spr_tutorial.toml
[2024-08-12 22:37:18,031]-[main] INFO  Backend - Reading workflow: /opt/nextPYP/workflows/tomo_tutorial.toml
[2024-08-12 22:37:18,057]-[main] INFO  Backend - Reading workflow: /opt/nextPYP/workflows/class_tutorial.toml
[2024-08-12 22:37:18,709]-[main] INFO  org.reflections.Reflections - Reflections took 610 ms to scan 1 urls, producing 116 keys and 6050 values
[2024-08-12 22:37:18,973]-[main] INFO  ktor.application - Responding at http://127.0.0.1:8080
```

**Comment by ccgauvin94** [comment]
One other thing of note, once I get this initial "Internal Server Error", it seems to cause cascading problems where any UI interaction just causes more "Internal Server Error", but again, I can't find any log output.

**Comment by cuchaz** [comment]
Hi there. Sorry you're running into trouble here. I'm not quite sure what the issue is here because some of the info you reported doesn't quite make sense to me yet. I hope you don't mind if I ask a few clarifying questions.

But first some background info. If the worker process is seeing "Internal Error" messages, that means there should be a more detailed error message on the web server `micromon` log. It's very strange that you're not seeing the error message there so something unexpected must be happening.

Are you by any chance running multiple different instances of the web server? Possibly in different locations?

The config file you posted shows the local folder at: `/cryoem_share/nextpyp/local` which means the log file for that web server should appear under `/cryoem_share/nextpyp/local/logs/micromon`.

However, your posted log file shows its configured local folder at: `/opt/nextPYP/local`. So the log file for that web server should be at `/opt/nextPYP/local/logs/micromon`.

Are there two separate log files in your environment? According to your config, the worker process should be trying to connect to the web server at `http://pipeline-20518:8081`. Which log file does that correspond to?

**Comment by ccgauvin94** [comment]
Hi @cuchaz - thanks so much for looking at this. I will gather that data and report back, but I'm also wondering: nextPYP needs to be installed/available on both the master, and worker, correct? If running two copies of the server is problematic, what prevents the server from running on the worker nodes? Do I need a separate image for the worker that does not run the SystemD process?

**Comment by cuchaz** [comment]
You should only need one web server running. The worker nodes won't need to run a web server. The worker processes will communicate with the web server over the network.

> Do I need a separate image for the worker that does not run the SystemD process?

Yes. The `container = '/opt/nextPYP/pyp.sif'` path points to the container image for the worker processes. There are no servers in there.

**Comment by ccgauvin94** [comment]
OK, I was able to significantly simplify things with that information. Here's how things are now laid out: 

Master:
* /opt/nextPYP contains config.toml and nextPYP.sif, has running systemd process for server. 
* mounts /cryoem_share/nextPYP

Worker:
* No running systemd process for server, mounts /cryoem_share/nextPYP

Shared filesystem:
* Contains local, shared, runtime/pip.sif, scratch, and workflows so both the server and worker will log to it. 

Here is my config.toml:
```
[pyp]
container = '/cryoem_share/nextPYP/runtime/pyp.sif'
scratch = '/cryoem_share/nextPYP/scratch'
binds = [ '/cryoem_share' ]
[web]
port = 8081
localDir = '/cryoem_share/nextPYP/local'
sharedDir = '/cryoem_share/nextPYP/shared'
workflowDirs = ['/cryoem_share/nextPYP/workflows']
auth = "none"
host = '0.0.0.0'
webhost = 'http://pipeline-20742:8081'
[slurm]
host = 'pipeline-20742'
queues = ['main.q']
```

Now, when I run a job, it spins up the container. Then it runs the job (apparently successfully). I can see the gain corrected images. However, the job never finishes. It just spins. 

Here is the output from /cryoem_share/nextPYP/shared/log/<the job log file>:

```
Fontconfig error: No writable cache directories
Fontconfig error: No writable cache directories
Fontconfig error: No writable cache directories
Matplotlib is building the font cache; this may take a moment.
2024-08-16 21:04:27 [INFO] /opt/pyp/bin/run/pyp:3232 | nextPYP v0.6.5 launching job 2 on pipeline-20743 using 1 task(s) and 4 GB of RAM
2024-08-16 21:04:28 [INFO] /opt/pyp/bin/run/pyp:380 | Attempting to load existing parameters and metadata from /cryoem_share/KOHLMA_GS_SUN_V1722941820_06262024/data/smaller_subset
2024-08-16 21:04:28 [INFO] /opt/pyp/bin/run/pyp:422 | /cryoem_share/KOHLMA_GS_SUN_V1722941820_06262024/data/smaller_subset/*.tif found 9 files to link into raw folder
2024-08-16 21:04:30 [INFO] /opt/pyp/bin/run/pyp:4098 | Selecting image for preview: /cryoem_share/<redacted>_V1722941820_06262024/data/smaller_subset/20240614_<redacted>-V1722941820_1_Jun
15_22.01.02.tif
2024-08-16 21:04:35 [INFO] /opt/pyp/bin/run/pyp:4101 | Image dimensions are 5760 x 4092 (37 frames/tilts)
2024-08-16 21:05:37 [INFO] /opt/pyp/bin/run/pyp:4123 | PYP (pypgain) finished successfully
Traceback (most recent call last):
  File "/opt/pyp/bin/run/webrpc", line 47, in <module>
    cli()
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/opt/pyp/bin/run/webrpc", line 23, in slurm_ended
    Web().slurm_ended(_arrayid(), exit)
  File "/opt/pyp/src/pyp/streampyp/web.py", line 78, in slurm_ended
    self._request(Request.slurm_ended(webid=self.webid, arrayid=arrayid, exit_code=exit_code))
  File "/opt/pyp/src/pyp/streampyp/web.py", line 39, in _request
    response = self.client.send(request)
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/apply_defaults/decorators.py", line 13, in wrapper
    return function(self, *args, **kwargs)
  File "/usr/local/envs/pyp/lib/python3.8/site-packages/jsonrpcclient/client.py", line 177, in send
    raise ReceivedErrorResponseError(response.data)
jsonrpcclient.exceptions.ReceivedErrorResponseError: Internal Error
```

Now there is a server log in /cryoem_share/nextPYP/local/logs/micromon:

```
[2024-08-16 21:05:59,377]-[eventLoopGroupProxy-4-7] ERROR JsonRpc.pyp - request: {"jsonrpc":"2.0","method":"slurm_ended","params":{"webid":"fM5dTlCvloWlYVN1","arrayid":null,"exit_code":0},"id":
1,"token":"AgmiDxY7NXqi1z7cZFeGjtlgagQ2l4UI4OqXAJ7anmkdl0Hn1vlfJYkF2xp5uF4PGR5oHN5NIlvb6M73ZD9XK8"}
com.jcraft.jsch.JSchException: channel is not opened.
        at com.jcraft.jsch.Channel.sendChannelOpen(Channel.java:772)
        at com.jcraft.jsch.Channel.connect(Channel.java:150)
        at com.jcraft.jsch.Channel.connect(Channel.java:144)
        at edu.duke.bartesaghi.micromon.SSHConnection$sftp$1.invoke(SSH.kt:337)
        at edu.duke.bartesaghi.micromon.SSHConnection$sftp$1.invoke(SSH.kt:335)
        at edu.duke.bartesaghi.micromon.SSHConnection.session(SSH.kt:275)
        at edu.duke.bartesaghi.micromon.SSHConnection.sftp(SSH.kt:335)
        at edu.duke.bartesaghi.micromon.cluster.slurm.SBatch$jobResult$out$1.invoke(SBatch.kt:220)
        at edu.duke.bartesaghi.micromon.cluster.slurm.SBatch$jobResult$out$1.invoke(SBatch.kt:219)
        at edu.duke.bartesaghi.micromon.SSHPool$connection$2.invokeSuspend(SSH.kt:103)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
[2024-08-16 21:05:59,381]-[eventLoopGroupProxy-4-7] WARN  Backend - Slow request: 20322 ms, uri=/pyp
```
It appears to be an SSH issue that occurred right after the job completed. Presumably, it can't communicate back to the web server that the job was complete? 

I opened up the worker, switched to the `nextpyp` user, and ssh'd into the master without issue, so my keys appear to be working. Not sure what elese might be the issue here as I can't quite discern what the log is saying, exactly.

**Comment by cuchaz** [comment]
Yes, that configuration makes a lot more sense to me. Nice job!

Looks like the error you're getting is the web server failing to process the end-of-job signal. The root cause seems to be an SSH connection failure. It looks like the worker was able to contact the web server correctly, but the web server wasn't able to log back into the SLURM login node to perform the post-job cleanup steps.

That error message is hopelessly vague though. It's coming from inside of the SSH library we're using, so we don't have much control over it. Digging into the SSH library code a bit, it looks like the SSH client wasn't able to establish the SFTP connection. Possibly because of a timeout? Or possibly because of an unfavorable response from the SSH daemon? Hard to say.

Hopefully it's just an intermittent network issue and it won't happen again? Since the job launched correctly (which also uses an SFTP connection), maybe just trying again wouldn't give the same error?

If that same error happens every time, that could mean your SSH daemon doesn't support our client somehow? Or it's deciding not to allow the SFTP connection somehow? We've seen SSH daemons that don't support SFTP in other places, so in the latest (not released yet) version of nextPYP, we've redesigned the system to cleanup jobs without using SFTP anymore. That might not help you too much right now, but there's hope that if this is a deterministic error, the next version of nextPYP might work much better there. =)

**Comment by ccgauvin94** [comment]
Thanks so much for the very speedy reply!

If I check, `sshd` is bound to port 22. `/etc/ssh/sshd_conf` seems to have sftp enabled: 

`Subsystem       sftp    /usr/libexec/openssh/sftp-server`

I will try sftp connection directly from the worker node.

**Comment by ccgauvin94** [comment]
SFTP is working from the worker node to the master node. 

I tried restarting everything, and lo and behold it now seems to be working. I submitted a number of jobs, which created new workers and ran successfully. So I'm going to close this and say the problem was potentially just that the worker nodes were also starting servers, that were then trying to write to the same DB and such. 

Thanks again for your help, really looking forward to getting going with nextPYP.
